cumulus/pallets/weight-reclaim/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+	const IDENTIFIER: &'static str = "StorageWeightReclaim<Use `metadata()`!>";
+
+	type Implicit = S::Implicit;
+
+	// Initial proof size and inner extension value.
+	type Val = (Option<u64>, S::Val);
+
+	// Initial proof size and inner extension pre.
+	type Pre = (Option<u64>, S::Pre);
+
+	fn implicit(&self) -> Result<Self::Implicit, TransactionValidityError> {
+		self.0.implicit()
+	}
+
+	fn metadata() -> Vec<sp_runtime::traits::TransactionExtensionMetadata> {
+		S::metadata()
+	}
+
+	fn weight(&self, call: &T::RuntimeCall) -> Weight {
+		T::WeightInfo::storage_weight_reclaim().saturating_add(self.0.weight(call))
+	}
+
+	fn validate(
+		&self,
+		origin: T::RuntimeOrigin,
+		call: &T::RuntimeCall,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		len: usize,
+		self_implicit: Self::Implicit,
+		inherited_implication: &impl Encode,
+	) -> Result<(ValidTransaction, Self::Val, T::RuntimeOrigin), TransactionValidityError> {
+		let proof_size = get_proof_size();
+
+		self.0
+			.validate(origin, call, info, len, self_implicit, inherited_implication)
+			.map(|(validity, val, origin)| (validity, (proof_size, val), origin))
+	}
+
+	fn prepare(
+		self,
+		val: Self::Val,
+		origin: &T::RuntimeOrigin,
+		call: &T::RuntimeCall,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let (proof_size, inner_val) = val;
+		self.0.prepare(inner_val, origin, call, info, len).map(|pre| (proof_size, pre))
+	}
+
+	fn post_dispatch_details(
+		pre: Self::Pre,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		post_info: &PostDispatchInfoOf<T::RuntimeCall>,
+		len: usize,
+		result: &DispatchResult,
+	) -> Result<Weight, TransactionValidityError> {
+		let (pre_dispatch_proof_size, inner_pre) = pre;
+
+		let mut post_info_with_inner = *post_info;
+		S::post_dispatch(inner_pre, info, &mut post_info_with_inner, len, result)?;
+
+		let inner_refund = if let (Some(before_weight), Some(after_weight)) =
+			(post_info.actual_weight, post_info_with_inner.actual_weight)
+		{
+			before_weight.saturating_sub(after_weight)
+		} else {
+			Weight::zero()
+		};
+
+		let Some(pre_dispatch_proof_size) = pre_dispatch_proof_size else {
+			// We have no proof size information, there is nothing we can do.
+			return Ok(inner_refund);
+		};
+
+		let Some(post_dispatch_proof_size) = get_proof_size() else {
+			log::debug!(
+				target: LOG_TARGET,
+				"Proof recording enabled during prepare, now disabled. This should not happen."
+			);
+			return Ok(inner_refund)
+		};
+
+		// The consumed proof size as measured by the host.
+		let measured_proof_size = post_dispatch_proof_size.saturating_sub(pre_dispatch_proof_size);
+
+		// The consumed weight as benchamrked.
+		// NOTE: `calc_actual_weight` will take the minimum of `post_info` and `info` weights.
+		// This means any underestimation of compute time in the pre dispatch info will not be
+		// taken into account.
+		let benchmarked_weight = post_info_with_inner.calc_actual_weight(info);
```

Comment:
I'd find it easier to read if this was `benchmarked_actual_weight` or something, the current name implies it's from before dispatch.

Reply:
Yes, I renamed and tried to keep the concept of "actual" https://github.com/paritytech/polkadot-sdk/pull/6140/commits/5322297303f165cc1f55ec32ac6779ec49a4bfbc